provider/block_dirty_bitmap.py

```diff
@@ -186,3 +186,58 @@ def get_bitmaps_in_device(vm, device):
     out = vm.monitor.cmd("query-block")
     bitmaps = get_bitmaps(out)
     return bitmaps.get(device, list())
+
+
+@fail_on
+def handle_block_dirty_bitmap_transaction(vm, disabled_params=None,
+                                          added_params=None,
+                                          merged_params=None):
+    """
+    Add/disable/merge bitmaps in one transaction.
+    :param vm: an active VM object
+    :param disabled_params: dict for bitmaps to be disabled,
+                  required: bitmap_device_node, bitmap_name
+                  optional: bitmap_disable_cmd
+    :param added_params: dict for bitmaps to be added,
+               required: bitmap_device_node, bitmap_name
+               optional: bitmap_add_cmd, bitmap_granularity,
+                         bitmap_persistent, bitmap_disabled
+    :param merged_params: dict for bitmaps to be merged
+                required: bitmap_device_node, bitmap_target, bitmap_sources
+                optional: bitmap_merge_cmd
+    """
+    actions = []
+
+    if disabled_params:
+        bitmap_disable_cmd = disabled_params.get('bitmap_disable_cmd',
+                                                 'block-dirty-bitmap-disable')
+        bitmap_data = {"node": disabled_params['bitmap_device_node'],
+                       "name": disabled_params['bitmap_name']}
+        actions.append({"type": bitmap_disable_cmd, "data": bitmap_data})
+
+    if added_params:
+        bitmap_add_cmd = added_params.get('bitmap_add_cmd',
+                                          'block-dirty-bitmap-add')
+        bitmap_data = {"node": added_params['bitmap_device_node'],
+                       "name": added_params['bitmap_name']}
+        if added_params.get('bitmap_granularity'):
+            bitmap_data['granularity'] = added_params['granularity']
+
+        mapping = {'on': True, 'yes': True, 'off': False, 'no': False}
+        if added_params.get('bitmap_persistent'):
+            bitmap_data['persistent'] = mapping[added_params['bitmap_persistent']]
+        if added_params.get('bitmap_disabled'):
+            bitmap_data['disabled'] = mapping[added_params['bitmap_disabled']]
+        actions.append({'type': bitmap_add_cmd, 'data': bitmap_data})
+
+    if merged_params:
+        bitmap_merge_cmd = merged_params.get('bitmap_merge_cmd',
+                                             'block-dirty-bitmap-merge')
+        bitmap_data = {'node': merged_params['bitmap_device_node'],
+                       'target': merged_params['bitmap_target'],
+                       'bitmaps': merged_params['bitmap_sources']}
+        actions.append({'type': bitmap_merge_cmd, 'data': bitmap_data})
+
+    if actions:
+        arguments = {"actions": actions}
+        vm.monitor.cmd("transaction", arguments)
```

Comment:
Suggest to split it into three API,  more clear, let's keep one function do one thing.
e.g:
1. `def add_bitmaps_in_singel_transaction(...)`
2. `def disable_bitmaps_in_singel_transaction(...)`
3. `def merge_bitmaps_in_singel_transaction(...)`

Then you could wrap them into a higher layer in case level, like :
https://github.com/autotest/tp-qemu/pull/2369/files#diff-ddfb1ce02cb896c1232aebbf9ef9c337
  ```
  def add_bitmaps_transaction(self):
        for i, bitmap in enumerate(self.full_backup_bitmaps):
            disabled_params = {'bitmap_device_node': self.source_images[i],
                               'bitmap_name': bitmap}
            added_params = {'bitmap_device_node': self.source_images[i],
                            'bitmap_name': self.before_2nd_inc_bitmaps[i]}
            block_dirty_bitmap.disable_bitmaps_in_singel_transaction(disabled_params...)
            block_dirty_bitmap.add_bitmaps_in_singel_transaction(added_params...)
```
But I don't know that split them whether to conform to your testing requirements.

Reply:
Please add the target file name into your commit header, like:
`block_dirty_bitmap: Handle bitmaps in one transaction`

Reply:
Thank you, @YongxueHong .
1. I am afraid we cannot split it into several functions, that's because we do all kinds of actions(disable/add/merge) in one and only one transaction, e.g. 
{'execute': 'transaction', 'arguments': {'actions': [{'type': 'block-dirty-bitmap-disable', 'data': {'node': 'drive_data', 'name': 'before_2nd_inc_bitmap_data'}}, {'type': 'block-dirty-bitmap-add', 'data': {'node': 'drive_data', 'name': 'merged_bitmap_data', 'disabled': True}}, {'type': 'block-dirty-bitmap-merge', 'data': {'node': 'drive_data', 'target': 'merged_bitmap_data', 'bitmaps': ['full_bitmap_data', 'before_2nd_inc_bitmap_data']}}]}, 'id': '88zuaWgm'}

2. updated
block_dirty_bitmap: Handle all bitmap actions in one transaction
This may introduce the fix more clearly